containers/flatpak/prepare

```diff
@@ -1,83 +1,212 @@
-
-ARCHIVE_LOCATION="${1:-}"
-ARCHIVE_SHA256="${2:-}"
-
-if test -z "${ARCHIVE_LOCATION}"; then
-    ARCHIVE_LOCATION="$("${0%/*}/../../tools/make-dist" IGNORE_UNUSED_PATTERNS=1 ONLYDIR=static)"
-fi
-
-case "${ARCHIVE_LOCATION}" in
-    http?://*)
-        ARCHIVE_TYPE="url"
-        FLATPAK_BRANCH=stable
-        ;;
-    /*)
-        ARCHIVE_TYPE="path"
-        FLATPAK_BRANCH=devel
-        ;;
-    *)
-        echo 'If specified, archive must be an https URL or an absolute local pathname' >&2
-        exit 1
-        ;;
-esac
-
-if test -z "${ARCHIVE_SHA256}"; then
-    if test -f "${ARCHIVE_LOCATION}"; then
-        ARCHIVE_SHA256="$(sha256sum "${ARCHIVE_LOCATION}" | cut -f1 -d' ')"
-    else
-        echo 'Archive must exist locally, or checksum must be explicitly given.' >&2
-        exit 1
-    fi
-fi
-
-printf "  %-8s %s\n" GEN "${FLATPAK_ID}.yml" >&2
-sed \
-    -e "s|@FLATPAK_ID@|${FLATPAK_ID}|" \
-    -e "s|@FLATPAK_BRANCH@|${FLATPAK_BRANCH}|" \
-    -e "s|@ARCHIVE_TYPE@|${ARCHIVE_TYPE}|" \
-    -e "s|@ARCHIVE_LOCATION@|${ARCHIVE_LOCATION}|" \
-    -e "s|@ARCHIVE_SHA256@|${ARCHIVE_SHA256}|" \
-    -e "s|@PYBRIDGE_ENABLE@|${PYBRIDGE_ENABLE:-disable}|" \
-    "${0%/*}/cockpit-client.yml.in" > "${FLATPAK_ID}.yml.tmp"
-
-if [ "${PYBRIDGE_ENABLE:-}" = 'enable' ]; then
-    printf "
-  - name: cockpit-podman
-    buildsystem: simple
-    build-commands:
-      - make install PREFIX=/app
-
-    sources:
-      - type: archive
-        url: %s
-        sha256: %s
-" $(cat "${0%/*}/beiboot-extra-packages") >> "${FLATPAK_ID}.yml.tmp"
-fi
-
-mv "${FLATPAK_ID}.yml.tmp" "${FLATPAK_ID}.yml"
-touch "${FLATPAK_ID}.releases.xml"
-echo "${FLATPAK_ID}.yml"
+#!/usr/bin/python3
+
+import argparse
+import contextlib
+import hashlib
+import json
+import logging
+import os
+import subprocess
+import sys
+import urllib.request
+from pathlib import Path
+from typing import Any, List, Optional, Sequence
+
+logger = logging.getLogger(__name__)
+
+FLATPAK_ID = 'org.cockpit_project.CockpitClient'
+
+# Local filenames
+MANIFEST_JSON = f'{FLATPAK_ID}.json'
+PACKAGES_JSON = f'{FLATPAK_ID}.packages.json'
+RELEASES_XML = f'{FLATPAK_ID}.releases.xml'
+
+# Constants related to extra packages
+UPSTREAM_REPOS = [
+    'cockpit-project/cockpit-machines',
+    'cockpit-project/cockpit-ostree',
+    'cockpit-project/cockpit-podman',
+]
+DOWNSTREAM_PACKAGES_URL = f'https://raw.githubusercontent.com/flathub/{FLATPAK_ID}/master/{PACKAGES_JSON}'
+
+FLATPAK_DIR = os.path.dirname(__file__)
+TOP_DIR = os.path.dirname(os.path.dirname(FLATPAK_DIR))
+
+
+def write_json(filename: str, content: Any) -> None:
+    """Something like g_file_set_contents() for JSON"""
+    tmpfile = f'{filename}.tmp'
+    try:
+        with open(tmpfile, 'w') as file:
+            json.dump(content, file, indent=2)
+    except BaseException:
+        with contextlib.suppress(OSError):
+            os.unlink(tmpfile)
+        raise
+    else:
+        os.rename(tmpfile, filename)
```

Comment:
Meh, I was hoping `pathlib.Path.write_text()` would do atomic updates, but it doesn't :cry:

Reply:
This was my first impulse as well, and indeed.... lame.